pkg/agent/validations.go

```diff
@@ -2,138 +2,123 @@ package agent
+					previousMessage: "",
+				}
+			}
+			// First time we see a validation ID
+			if !validationHistory[r.ID].seen {
+				validationHistory[r.ID].currentMessage = r.Message
+				validationHistory[r.ID].currentStatus = r.Status
+			} else {
+				validationHistory[r.ID].previousMessage = validationHistory[r.ID].currentMessage
+				validationHistory[r.ID].previousStatus = validationHistory[r.ID].currentStatus
+				validationHistory[r.ID].currentMessage = r.Message
+				validationHistory[r.ID].currentStatus = r.Status
+			}
 			switch r.Status {
 			case validationFailure, validationError:
-				label := r.ID
-				if v, ok := hostValidationLabels[r.ID]; ok {
-					label = v
-				}
-
-				traces = append(traces, validationTrace{
-					header:   errorMsg,
-					category: category,
-					label:    label,
-					message:  r.Message,
-				})
+				validationHistory[r.ID].numFailures++
 			}
+			logValidationHistory(logPrefix, validationHistory[r.ID], log)
 		}
 	}
-
-	return traces
+	return validationHistory, nil
 }
 
-func checkHostsValidations(cluster *models.Cluster, log *logrus.Logger) bool {
-
-	var currentValidations []validationTrace
-
-	currentValidations = append(currentValidations, logValidationsStatus("Validation failure found for cluster", cluster.ValidationsInfo, log)...)
-	for _, h := range cluster.Hosts {
-		currentValidations = append(currentValidations, logValidationsStatus(fmt.Sprintf("Validation failure found for %s", h.RequestedHostname), h.ValidationsInfo, log)...)
-	}
-
-	sort.Slice(currentValidations, func(i, j int) bool {
-		if currentValidations[i].header != currentValidations[j].header {
-			return currentValidations[i].header < currentValidations[j].header
-		}
-		if currentValidations[i].category != currentValidations[j].category {
-			return currentValidations[i].category < currentValidations[j].category
-		}
-		return currentValidations[i].label < currentValidations[j].label
-	})
-
-	if !reflect.DeepEqual(currentValidations, previousValidations) {
-		previousValidations = currentValidations
-
-		if len(previousValidations) == 0 {
-			log.Info("Pre-installation validations are OK")
-			return true
+func logValidationHistory(logPrefix string, history *validationResultHistory, log *logrus.Logger) string {
+	loglevel := ""
+	// Fist time we print anything
+	if !history.seen {
+		switch history.currentStatus {
+		case validationFailure, validationError:
+			log.Warning(logPrefix + history.currentMessage)
+			loglevel = "warning"
+		case validationSuccess:
+			log.Debug(logPrefix + history.currentMessage)
+			loglevel = "debug"
 		}
+		history.seen = true
+		return loglevel
+	}
 
-		log.Debug("Checking for validation failures ----------------------------------------------")
-		for _, v := range previousValidations {
-			log.WithFields(logrus.Fields{
-				"category": v.category,
-				"label":    v.label,
-				"message":  v.message,
-			}).Debug(v.header)
+	if history.currentStatus != history.previousStatus {
```

Comment:
This check means that if the message changes but the status stays the same, we won't log anything. I'm not sure if this is a problem in practice, but if it happens I would have assumed we'd want to print the new message again. The algorithm I had in mind was quite simple:

```
if history.currentMessage != history.PreviousMessage {
    switch history.currentStatus {
    case validationFailure, validationError:
        log.Warning(...)
    case validationSuccess {
        if history.PreviousMessage != "" {
            log.Info(...)
        } else {
            log.Debug(...)
        }
    }
}
```

Reply:
Changed to look for difference in message vs status